Library/Homebrew/utils/lock.sh

```diff
@@ -0,0 +1,31 @@
+# create a lock using `flock (2)`. A name is required as first argument.
+# the lock will be automatically unlocked when the shell process quits.
+lock() {
+  local name="$1"
+  local lock_file="/tmp/homebrew-${name}.lock"
+  # open the lock file to file descriptor 3, so the shell process can hold the lock.
+  exec 3>"$lock_file"
+  if ! _create_lock
+  then
+    odie <<EOS
+Another active Homebrew process is already in process.
+Please wait for it to finish or terminate it to continue.
+EOS
+  fi
+}
+
+_create_lock() {
+  if [[ -x /usr/bin/ruby ]]
+  then
+    /usr/bin/ruby -e 'exit 1 unless File.new(3).flock(File::LOCK_EX|File::LOCK_NB)'
+  elif [[ -x /usr/bin/flock ]]
+  then
+    /usr/bin/flock -n 3
+  elif [[ -x /usr/bin/python ]]
+  then
+    /usr/bin/python -c 'import os, fcntl; fcntl.flock(os.fdopen(3, "w"), fcntl.LOCK_EX | fcntl.LOCK_NB)' 2>/dev/null
```

Comment:
This doesn't introduce a Python dependency. It is just another alternative if you don't have ruby and flock. But that to be said, I'm fine to remove this case.